### PR TITLE
Fix nil error for empty files in tars

### LINKED ISSUE #4616: inspec json fails for tar archives with empty files
There is a blank file in this dev-sec profile:
https://github.com/dev-sec/cis-docker-benchmark/blob/master/libraries/.gitkeep

It probably shouldn't be there, but it's recently started failing `inspec json` due to the fix we did for character encoding with tar files.

NOTE: (@clintoncwolfe): The profile here should be archived using `inspec archive`, then the repro invocation is `inspec json profile-1.2.3.tar.gz` - you are inflating and exporting the profile as JSON from a tarball.

```

Generate inspec json for cis-docker-benchmark-2.1.0.tar.gz ...
--
  | Traceback (most recent call last):
  | 17: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-bin-4.18.0/bin/inspec:11:in `<main>'
  | 16: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/base_cli.rb:33:in `start'
  | 15: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
  | 14: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
  | 13: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
  | 12: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
  | 11: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/cli.rb:79:in `json'
  | 10: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/profile.rb:73:in `for_target'
  | 9: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/profile.rb:67:in `for_fetcher'
  | 8: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/profile.rb:49:in `for_path'
  | 7: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/file_provider.rb:14:in `for_path'
  | 6: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/file_provider.rb:14:in `new'
  | 5: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/file_provider.rb:160:in `initialize'
  | 4: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/file_provider.rb:205:in `walk_tar'
  | 3: from /hab/pkgs/core/ruby/2.5.7/20191007205439/lib/ruby/site_ruby/2.5.0/rubygems/package/tar_reader.rb:29:in `new'
  | 2: from /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/file_provider.rb:161:in `block in initialize'
  | 1: from /hab/pkgs/core/ruby/2.5.7/20191007205439/lib/ruby/site_ruby/2.5.0/rubygems/package/tar_reader.rb:65:in `each'
  | /hab/pkgs/chef/inspec/4.18.0/20191010204024/lib/gems/inspec-4.18.0/lib/inspec/file_provider.rb:176:in `block (2 levels) in initialize': undefined method `force_encoding' for nil:NilClass (NoMethodError)


```

## Possible Solution
I opened PR https://github.com/inspec/inspec/pull/4615

### PR DESCRIPTION
**Pull request by james-stocks**
entry.read returns nil when entry has size 0

Signed-off-by: James Stocks <jstocks@chef.io>

edit - Fixes #4616

### CODE REVIEW

File: lib/inspec/file_provider.rb [hunk 1]
```diff
@@ -172,9 +172,11 @@ def initialize(path)
 
           @contents[path] = begin # not ||= in a tarball, last one wins
                               res = entry.read
-                              try = res.dup
-                              try.force_encoding Encoding::UTF_8
-                              res = try if try.valid_encoding?
+                              if(res) # not an empty file
```
Comment (james-stocks) [inline lib/inspec/file_provider.rb]: I was initially going to skip this entry at line 166, but I think we want to respect that the file is in the @contents even if it's empty?

**Review by clintoncwolfe — DISMISSED** [review]
I think the implementation change is probably OK, but this change should be tested - I would like to see a functional test on this.

**Comment by clintoncwolfe** [comment]
I'm recusing myself - I've contributed the tests and repro.

File: test/functional/inspec_json_profile_test.rb [hunk 2]
```diff
@@ -178,4 +178,18 @@
...
   end
+
+  # regression #4615
+  describe "inspec json with an archived profile containing an empty file" do
+    it "does not crash" do
+      out = inspec("json " + File.join(profile_path, "contains-empty-file-0.1.0.tar.gz"))
+
+      assert_equal "contains-empty-file", JSON.load(out.stdout)["name"]
+
+      _(out.stderr).must_equal ""
```
Comment (miah) [inline test/functional/inspec_json_profile_test.rb]: I think you can swap `must_equal` for `must_be_empty`

File: test/unit/mock/profiles/contains-empty-file/controls/example.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# copyright: 2018, The Authors
```
Comment (miah) [inline test/unit/mock/profiles/contains-empty-file/controls/example.rb:1]: Do we need a valid control in the test profile to trigger this?

**Review by miah — CHANGES_REQUESTED** [review]
Can you take a look at my inline comments? Thanks.

File: test/unit/mock/profiles/contains-empty-file/controls/example.rb (hunk 3 above)
Comment (clintoncwolfe) [reply test/unit/mock/profiles/contains-empty-file/controls/example.rb:1]: No, on master the test profile triggered the problem.

File: lib/inspec/file_provider.rb [hunk 4]
```diff
@@ -172,9 +172,11 @@ def initialize(path)
 
           @contents[path] = begin # not ||= in a tarball, last one wins
                               res = entry.read
-                              try = res.dup
-                              try.force_encoding Encoding::UTF_8
-                              res = try if try.valid_encoding?
+                              if (res) # not an empty file
```
Comment (zenspider) [inline lib/inspec/file_provider.rb:175]: Not C. No parens pls

File: test/unit/mock/profiles/contains-empty-file/controls/example.rb (hunk 3 above)
Comment (zenspider) [reply test/unit/mock/profiles/contains-empty-file/controls/example.rb:1]: Why does this PR include a whole profile directory when this is about the tarball?

File: test/functional/inspec_json_profile_test.rb [hunk 5]
```diff
@@ -178,4 +178,18 @@
...
+  describe "inspec json with an archived profile containing an empty file" do
+    it "does not crash" do
+      out = inspec("json " + File.join(profile_path, "contains-empty-file-0.1.0.tar.gz"))
+
+      assert_equal "contains-empty-file", JSON.load(out.stdout)["name"]
+
+      _(out.stderr).must_be_empty ""
+
+      assert_exit_code 0, out
+    end
```
Comment (zenspider) [inline test/functional/inspec_json_profile_test.rb:192]: I think this is overkill. We don't need a whole integration test to verify that we handle tarballs with empty files correctly. I'm going to work out an alternative because this should be a 1 line impl change, and very short (actual) functional test, and your tarball.

**Review by zenspider — CHANGES_REQUESTED** [review]
I'd like to push back on this one. Gimme a few minutes for an alternative.

**Comment by zenspider** [comment]
See #4619 as my counter PR.

**Comment by zenspider** [comment]
Fixed in #4619. Closing.
